lightkube/utils/quantity.py

```diff
@@ -0,0 +1,142 @@
+
+    **returns**  An instance of `decimal.Decimal` representing the quantity as a bare decimal.
+    """
+    if quantity is None:
+        # This is useful for comparing e.g. ResourceRequirements.limits.get("cpu"), which can be
+        # None.
+        return None
+
+    pat = re.compile(r"([+-]?\d+(?:[.]\d*)?(?:e[+-]?\d+)?|[.]\d+(?:e[+-]?\d+)?)(.*)")
+    match = pat.match(quantity)
+
+    if not match:
+        raise ValueError("Invalid quantity string: '{}'".format(quantity))
+
+    try:
+        value = decimal.Decimal(match.group(1))
+    except ArithmeticError as e:
+        raise ValueError("Invalid numerical value") from e
+
+    unit = match.group(2)
+
+    try:
+        multiplier = MULTIPLIERS[unit]
+    except KeyError:
+        raise ValueError("Invalid unit suffix: {}".format(unit))
+
+    try:
+        as_decimal = value * multiplier
+        return as_decimal.quantize(decimal.Decimal("0.001"), rounding=decimal.ROUND_UP)
+    except ArithmeticError as e:
+        raise ValueError("Invalid numerical value") from e
+
+
+def _equals_canonically(first_dict: Optional[dict], second_dict: Optional[dict]) -> bool:
+    """Compare resource dicts such as 'limits' or 'requests'."""
+    if first_dict == second_dict:
+        # This covers two cases: (1) both args are None; (2) both args are identical dicts.
+        return True
+    if first_dict and second_dict:
+        if first_dict.keys() != second_dict.keys():
+            # The dicts have different keys, so they cannot possibly be equal
+            return False
+        return all(
+            parse_quantity(first_dict[k]) == parse_quantity(second_dict[k])
+            for k in first_dict.keys()
+        )
+    if not first_dict and not second_dict:
+        # This covers cases such as first=None and second={}
+        return True
+    return False
+
+
+@overload
+def equals_canonically(first: ResourceRequirements, second: ResourceRequirements) -> bool:
+    ...
+
+
+@overload
+def equals_canonically(first: Optional[dict], second: Optional[dict]) -> bool:
+    ...
+
+
+def equals_canonically(first, second):
+    """Compare two resource requirements for numerical equality.
+
+    Both arguments must be of the same type and can be either:
+    - `ResourceRequirements`; or
+    - Optional[dict], representing the "limits" or the "requests" portion of `ResourceRequirements`.
+
+    >>> equals_canonically({"cpu": "0.6"}, {"cpu": "600m"})
+    True
+    >>> equals_canonically(ResourceRequirements(limits={"cpu": "0.6"}), ResourceRequirements(limits={"cpu": "600m"}))
+    True
+
+    **Parameters**
+
+    * **first** `ResourceRequirements` or `dict` - The first item to compare.
+    * **second** `ResourceRequirements` or `dict` - The second item to compare.
+
+    **returns**  True, if both arguments are numerically equal; False otherwise.
+    """
+    if isinstance(first, Optional[dict]) and isinstance(second, Optional[dict]):
```

Comment:
Runtime error as typing constructs (`Optional`) cannot be used here.

Reply:
Didn't show up when running tests locally with `platform linux -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0`.
Fixed in next commit.